### DISCUSSION TITLE
`ts-rs` and types from external crates

### DISCUSSION THREAD

**Discussion by escritorio-gustavo** [Aleph-Alpha/ts-rs#219] [General]
We are talking about how to handle types from external crates (from `crates.io` or other sources) in two PRs (#211 and #218), so I thought this would be easier to follow

**Comment by NyxCode** [comment]
Good idea, we should probably make more use of the discussions feature!

So yeah, as I said, I can definetely see cases where it'd be nice for libraries to implement/derive `TS`, maybe without `#[ts(export)]`ing.  
Same goes for git or path dependencies for those who split their project into multiple crates, but aren't using a workspace.

**Reply by NyxCode** [reply]
Also, if we properly supported this, the workspace-situation might be solved with it. We might not even have to be workspace-aware if this just worked.

**Reply by escritorio-gustavo** [reply]
I currently have a working solution for workspaces, but I'm waiting on https://github.com/Aleph-Alpha/ts-rs/pull/211#issuecomment-1919294139 to commit it

**Reply by NyxCode** [reply]
That's good to hear. Still, I think it might make sense to first figure out in which direction we want to go with this issue here, since im not sure how the two will interact.

**Reply by escritorio-gustavo** [reply]
The reason I think it's weird to have external libs implementing `TS` is that it assumes the consumers need it, which is a big assumption because:
- The consumer must be using TypeScript in their project
- The consumer must need to translate between Rust and TS types
- The consumer wants to do this through `ts-rs`

All of these assumptions are a bit of a stretch for an external crate

If all of these are wrong, the library must then assume the consumer doesn't mind an extra dependency on `ts-rs`

Of course, all of these problems can be mitigated by hiding `ts-rs` behind a feature flag

**Reply by escritorio-gustavo** [reply]
Something we could try is to force external crate types to be `#[ts(inline)]`d, this way we avoid the `#[ts(export)]` problem

**Reply by escritorio-gustavo** [reply]
I think I see your point on how it could be useful though, the same way a bunch of crates have a `serde` flag

**Comment by escritorio-gustavo** [comment]
@NyxCode can #195 somehow help with this?

**Reply by escritorio-gustavo** [reply]
Maybe external crates should use `#[ts(export)]` and the default directory is where the bindings go?

**Reply by escritorio-gustavo** [reply]
If this were to work, we may want to force `pub` `TS` types to use `#[ts(export)]`, but only if there is a `lib.rs` instead of `main.rs`? (is there a way to check this?)

**Reply by NyxCode** [reply]
Uhh, that's an interesting idea! Configure the output directory globally, and then every `.ts` file in the project and dependency tree goes there. That'd certainly work!

**Reply by escritorio-gustavo** [reply]
Okay, if we were to go with this, we'd have to find ways to enforce it.

> If this were to work, we may want to force pub TS types to use #[ts(export)], but only if there is a lib.rs instead of main.rs? (is there a way to check this?)

This (if it even is possible) would take care of the library side, but in the consumer side we'd have to enforce setting the environment variable (SQLx does this for its `query!` macros with the `DATABASE_URL` variable, which must be in a `.env` file in the project root)

**Reply by escritorio-gustavo** [reply]
Also, this configuration would have to be set for the whole workspace

**Reply by NyxCode** [reply]
> Okay, if we were to go with this, we'd have to find ways to enforce it.  

We could, like sqlx, introduce a CLI tool, and you'd export stuff with `ts-rs export ./bindings`, for example.  
This could set the environment variables[^1] and then run `TS_EXPORT=1 cargo test`. 

[^1]: With this approach, we could also introduce a config file! The CLI tool could read that config file, and convert it to environment variables, and then run `cargo t`.

**Reply by NyxCode** [reply]
I think I'm really warming up to the idea of introducing a CLI tool for this. Everyone already thinks the `cargo test` workaround is akward, so that might be a good reason to hide it behind a CLI tool. 

Edit: After playing with this a bit, I realize that a CLI really makes everything possible. It could get all dependencies with `cargo metadata`, and run their tests with `cargo t -p {dependency}` to generate the output.
Or the tests could write a temporary file, which could include their dependencies. Then the CLI could read those, and stitch everything together.

**Reply by escritorio-gustavo** [reply]
I like the CLI idea too, though i think `TS_EXPORT=1 cargo test` won't work on Windows, we'll have to find a way around that. Probably with `dotenv`, the same way sqlx does it

**Reply by escritorio-gustavo** [reply]
Alternatively, we could have the CLI add a feature flag like `ts-export` to the user's crate, gate our tests behind it and run `cargo test --features "ts-export <other_ts_rs_features_here>"`

**Reply by escritorio-gustavo** [reply]
Whatever we choose to do, will prevent `ts-rs`'s tests from running when the user just runs `cargo test`, but should we also try to prevent the user's tests from running when we run the CLI? (maybe filtering by tests named `export_bindings_*`)

**Reply by escritorio-gustavo** [reply]
Another point: types from external libraries that contain `#[derive(TS)]` but don't contain `#[ts(export)]` will now have to be exported. But what about types from the user's crate without `#[ts(export)]`?

**Reply by escritorio-gustavo** [reply]
> The "naive" approach of just exporting all dependencies with the type

This is a good idea when using the CLI, since it could handle deduping and avoid the corruption problem, we just gotta find a way to export from `Dependency`

**Reply by escritorio-gustavo** [reply]
Now a new problem is that #54 might become a big deal, since external crates could make naming collisions more common

**Reply by NyxCode** [reply]
> This is a good idea when using the CLI, since it could handle deduping and avoid the corruption problem, we just gotta find a way to export from Dependency

Precisely! I think I have an idea on how to implement it (having `TS::dependencies` return `impl TypeList`, where `() impl TypeList` and `(L: TypeList, PhantomData<T>: TS) impl TypeList` ). That's some type-level hackery, but i think it might work

**Reply by escritorio-gustavo** [reply]
That's interesting, and what would `TypeList` contain? Is it just an empty trait to give us a "reference" to `T`?

**Reply by escritorio-gustavo** [reply]
Also, check out [this comment](https://github.com/Aleph-Alpha/ts-rs/discussions/219#discussioncomment-8334541) this could be a simpler solution

**Reply by NyxCode** [reply]
> That's interesting, and what would TypeList contain? Is it just an empty trait to give us a "reference" to T?

Yeah, so the general idea is to build up a big generic type, where the generics store the types.  
A `TypeList` with the elements `i32, String, u32` would have the type `((((), (PhantomData<i32>,)), (PhantomData<String>,)), (PhantomData<u32>,))`.

Here's a quick demo: [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=17cf0c2ad7910c1de66f9dfb45049a9f)

I've got something similar working already, but I'm running into some `Sized` restrictions right now, which I'm not sure yet how to overcome.

**Reply by NyxCode** [reply]
For example, this is what 
```rust
    #[derive(TS)]
    #[allow(dead_code)]
    struct FullOfRefs<'a> {
        str_slice: &'a str,
        ref_slice: &'a [&'a str],
        num_ref: &'a i32,
    }
```
in `tests/references.rs` expands to:
```rust
fn dependencies2() -> impl ts_rs::typelist::TypeList
where
    Self: 'static,
{
    {
        use ts_rs::typelist::TypeList;
        ().push::<str>().push::<Vec<&'a str>>().push::<str>().push::<i32>()
    }
}
```
The `.push::<str>()` is a problem right now, since I had to introduce a `T: Sized` bound to `TypeList::push`

**Reply by escritorio-gustavo** [reply]
> `((((), (PhantomData<i32>,)), (PhantomData<String>,)), (PhantomData<u32>,))`

So this is almost like a linked list of `TypeList`?

**Reply by NyxCode** [reply]
yes!

**Reply by escritorio-gustavo** [reply]
> I've got something similar working already, but I'm running into some Sized restrictions right now, which I'm not sure yet how to overcome.

I suspect you have a more complex situation than the playground you linked, still, in that playground I just spammed `?Sized` until the errors went away
[Playgrround](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=ad66dd179eccd339beb305e1e03ce318)

**Reply by NyxCode** [reply]
> I suspect you have a more complex situation than the playground you linked, still, in that playground I just spammed ?Sized until the errors went away

That was the solution, btw., even though I dont completely understand why ^^

**Reply by escritorio-gustavo** [reply]
I think it's because we never get an actual value of type `T`, so it doesn't matter that we don't know its size

**Comment by NyxCode** [comment]
The "naive" approach of just exporting all dependencies with the type is the following:
Let's say we've got `A` and `B`, which contain an external type of a library:

```rust
#[ts(export)]
struct A {
  y: crate_y::Y	
}

#[ts(export)]
struct B {
  y: crate_y::Y	
}
````

For this, we'd expect this output:
```
bindings/
  A.ts
  B.ts
  Y.ts
```
`Y.ts` gets automatically exported as well, since the `#[ts(export)]`ed types depend on it.

However, the generated test for `B` doesn't know about the generated test for `A`, so both tests would try to write `Y.ts`. 
Here, we'd need some kind of locking (either a Mutex somewhere, or locking using the filesystem) to prevent file corruption.

**Reply by escritorio-gustavo** [reply]
If `crate_y::Y` had `#[ts(export)]`, would it work?

**Reply by escritorio-gustavo** [reply]
That way the tests for `A` and `B` don't have to export `Y` themselves

**Reply by escritorio-gustavo** [reply]
> If `crate_y::Y` had `#[ts(export)]`, would it work?

Nvm, tried it and it doesn't

**Reply by NyxCode** [reply]
I don't think so, since `cargo t` in the crate will not run the tests of `crate_y`. There is a way to do that (`cargo t -p crate_y`), but then, the output would probably land somewhere in `~/.cargo`, I think.

**Reply by escritorio-gustavo** [reply]
I ran `cargo test -p dependency1` from inside the `consumer` directory in `e2e` and it generated the bindings inside `dependency1`

**Reply by NyxCode** [reply]
Right - in that case, the library is actually in that directory. If it was a git or crates.io dependency, the library would be, as far as i know, in `~/.cargo`.

**Reply by escritorio-gustavo** [reply]
Maybe with the env variable proposed in #195 we could override this behavior?

**Reply by escritorio-gustavo** [reply]
> Maybe with the env variable proposed in https://github.com/Aleph-Alpha/ts-rs/pull/195 we could override this behavior?

This works, but requires the following

1. Create the file: `<project-root>/.cargo/config.toml`
2. Add the following to the file:
```toml
[env]
TS_RS_EXPORT_DIR = { value = "<RELATIVE_PATH_FROM_PROJECT_ROOT_TO_BINDINGS_DIR>", relative = true }
```
3. Run `cargo test -p <dependency>` 
4. Run `cargo test`

**Reply by escritorio-gustavo** [reply]
Also note that step 3 only needs to be done after `dependency` updates

**Reply by escritorio-gustavo** [reply]
Also, as a bonus, library authors can enforce the setting of this variable with this:
```rs
#[test]
fn env_set() {
    assert_ne!(env!("TS_RS_EXPORT_DIR"), "");
}
```

`cargo test -p <dependency>` will panic if the variable isn't set

**Reply by escritorio-gustavo** [reply]
Check out #220 to see how it works

**Comment by NyxCode** [comment]
# Possible Solution Nr.3
We re-introduce something like [export!](https://docs.rs/ts-rs/5.0.0/ts_rs/macro.export.html), and make exporting much more explicit (and centralized in one place!).

Users would have to explicitly name everything they want to export - including types their types contains.
```rs
export!(
  MyTypeA, MyTypeB<()> => "my_types.ts",
  SomeLibraryType => "lib.ts"
);
```

I got rid of that quite a while back and replaced it with `#[ts(export)]`, since I didn't like that you'd have to name every type you wanted to export. For example, every exported type would need to be public so that you can access it in your `main.rs`.

The big advantage of this here is that you could export multiple types in one file. This isn't easily possible with the current "one test per type" approach (though it might be, with file locking, idk)

**Reply by NyxCode** [reply]
In the workspace scenario, only the top-level crate(s) should then call `export!`, and include their dependencies in that.

**Reply by escritorio-gustavo** [reply]
I think i like the CLI better. As you mentioned, this forces the user to make everything `pub` and they need to know that passing `()` to the generic param is the proper way to do it

**Comment by NyxCode** [comment]
@escritorio-gustavo Now that we've explored two possible solutions already, what do you think?

Personally, I think I prefer the `TypeList` based one. It doesn't require a CLI right away (thogh I really want one), and from the perspective of a user everything just works like before.
My judgement might be clouded by the effort i put into that solution though, so I'd love to hear your opinion.

**Reply by escritorio-gustavo** [reply]
I like the `TypeList` a lot as well, especially since it ensures only types used by the consumer will get exported

**Reply by escritorio-gustavo** [reply]
It's definetly gonna solve #111 and #177 ata the same time, which is great!
It doesn't seem to clash with #195 at all, so that can be merged afterwards to allow setting a different default directory.
I don't think it solves #214, so #217 may still be needed

**Reply by NyxCode** [reply]
> It's definetly gonna solve https://github.com/Aleph-Alpha/ts-rs/issues/111 and https://github.com/Aleph-Alpha/ts-rs/discussions/177 ata the same time, which is great!

🥳 

> It doesn't seem to clash with https://github.com/Aleph-Alpha/ts-rs/pull/195 at all, so that can be merged afterwards to allow setting a different default directory.

Agreed, that's completely orthogonal

> I don't think it solves https://github.com/Aleph-Alpha/ts-rs/issues/214, so https://github.com/Aleph-Alpha/ts-rs/pull/217 may still be needed

Agreed! That might have been solved if we went down the `export! { .. }` route (maybe).
